standalone/pruntime/enclave/src/system/mod.rs

```diff
@@ -411,66 +421,260 @@ impl System {
+                        "Master key mismatches"
+                    );
+                    assert!(
+                        self.gatekeeper.is_none(),
+                        "Duplicated gatekeeper initialization"
+                    );
+
+                    info!("Init gatekeeper");
+                    let mut gatekeeper = gk::Gatekeeper::new(
+                        master_key.clone(),
+                        &mut block.recv_mq,
+                        self.send_mq
+                            .channel(MessageOrigin::Gatekeeper, master_key.clone()),
+                    );
+                    if self.registered_on_chain {
+                        gatekeeper.register_on_chain();
+                    }
+                    self.gatekeeper = Some(gatekeeper);
+                }
+            }
+        }
+    }
+
+    /// Monitor the getakeeper registeration event to:
+    ///
+    /// 1. Generate the master key if this is the first gatekeeper;
+    /// 2. Dispatch the master key to newly-registered gatekeepers;
+    fn process_new_gatekeeper_event(
+        &mut self,
+        block: &BlockInfo,
+        origin: MessageOrigin,
+        event: NewGatekeeperEvent,
+    ) {
+        if !origin.is_pallet() {
+            error!("Invalid origin {:?} sent a {:?}", origin, event);
+            return;
+        }
+
+        // double check the new gatekeeper is valid on chain
+        if !chain_state::is_gatekeeper(&event.pubkey, block.storage) {
+            error!("Fatal error: Invalid gatekeeper registration {:?}", event);
+            panic!("System state poisoned");
+        }
+
+        let my_pubkey = self.identity_key.public();
+        // gatekeeper genesis period
+        if event.gatekeeper_count == 1 {
+            if my_pubkey == event.pubkey && self.master_key.is_none() {
+                info!("Gatekeeper: generate master key as the first gatekeeper");
+                // generate master key as the first gatekeeper
+                // no need to restart
+                self.set_master_key(crate::new_sr25519_key(), false);
+            }
+
+            // upload the master key on chain
+            // noted that every gk should execute this to ensure the state consistency of egress seq
+            if let Some(master_key) = &self.master_key {
+                info!(
+                    "Gatekeeper: upload master key {} on chain",
+                    hex::encode(master_key.public())
+                );
+                let master_pubkey = RegistryEvent::MasterPubkey {
+                    master_pubkey: master_key.public(),
+                };
+                self.egress.send(&master_pubkey);
+            }
+        } else {
+            // dispatch the master key to the newly-registered gatekeeper using master key
+            // if this pRuntime is the newly-registered gatekeeper himself,
+            // its egress is still in dummy mode since we will tick the state later
```

Comment:
Each GK will dispatch a master key over the blockchain.
Can we move this branch of code into the Gatekeeper?  If we do so, we should also create the Gatekeeper instance on `NewGatekeeperEvent` rather than `MasterPubkeyOnChain`.

Reply:
Left a TODO